[hoo] Add with_effects to handle side effectful ops

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120296

Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side

angelayi committed:
[hoo] Add with_effects to handle side effectful ops

Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Result with ScriptObject:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=2] = placeholder[target=arg1_1]
    %_tensor_constant0 : [num_users=1] = get_attr[target=_tensor_constant0]
    %with_effects : [num_users=2] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, _TorchScriptTesting.takes_foo.default, %_tensor_constant0, %arg1_1), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 1), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %getitem_1), kwargs = {})
    return [getitem, add]
```

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,158 @@
+}
+
+
+with_effects = HigherOrderOperator("with_effects")
```
don't need this?

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -220,6 +220,12 @@ def __init__(self, pre_dispatch=False, export=False):
         self._mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
         # This will be turned off later for pre-dispatch functionalization
         self._dispatch_key = torch._C.DispatchKey.PreDispatch if pre_dispatch else None  # type: ignore[attr-defined]
+        # Map of functions or ScriptObjects to a token. The tokens help keep
+        # track of the ordering between side effectful operations. We have a
+        # token per ScriptObject and a token per function.
+        self._tokens: Dict[
+            Union[torch._ops.OpOverload, torch.ScriptObject], torch.Tensor
+        ] = {}
```
It might be overkill, but if we're going down the route of multiple effects, it's more that we should have "effects" and each "effect" should map to a token. Certain groups of ops may share the same effect (e.g. a "printing" effect). So this should be a dict["effect", torch.Tensor].

For starters, we can have one effect ("Ordered"?), and it's probably fine for it to be a string (could also be an enum).

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -332,6 +338,15 @@ def unwrap(x):
+
+        if has_effects(
+            func, args, kwargs
+        ) and not torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.Functionalize
+        ):
```
maybe just assert that there isn't a DispatchKey.Functionalize implementation

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,158 @@
+    unwrapped_kwargs = ctx.unwrap_tensors(kwargs)  # type: ignore[arg-type]
+    with ctx.redispatch_to_next():
+        (new_token, *unwrapped_outs) = with_effects(
+            unwrapped_token, op, *unwrapped_args, **unwrapped_kwargs  # type: ignore[arg-type]
```
there's a fun case of what happens if `op` is mutable or has alias information in addition to these new "effects". this will probably involve a with_effects wrapping an auto_functionalized HOP, which is fine. We should error out if `op` has any alias information at all.

angelayi replied:
I added an assertion at the definition of WithEffects for if the op has aliasing. Will leave the auto_functionalization as a followup.

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,158 @@
+
+    # Add the newly created token into the tokens map for a following call to
+    # use this token.
+    tokens[key] = new_token
```
we unwrapped tokens[key] above,  implying it was already wrapped. however, the new_token is unwrapped. What's going on?

angelayi replied:
😅 you're right. I wrapped it and put it back.

zou3519 commented on torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:
```diff
@@ -34,9 +33,11 @@
-    with enable_python_dispatcher(), FunctionalTensorMode(
-        pre_dispatch=aot_config.pre_dispatch, export=aot_config.is_export
-    ):
+    # with enable_python_dispatcher(), FunctionalTensorMode(
+    #     pre_dispatch=aot_config.pre_dispatch, export=aot_config.is_export
+    # ):
+    # HACK: checking with @bdhirsh
+    with enable_python_dispatcher():
```
What's going on here?

angelayi replied:
Updated the code here!

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,85 @@
+            "torch._higher_order_ops.effects.with_effects(arg0_1, torch.ops.aten.print.default, 'moo')", 1, exactly=True
+        ).run(gm.code)
+        self.assertEqual(len(gs.input_tokens), 1)
+        self.assertEqual(len(gs.output_tokens), 1)
```
Can you do a self.assertExpectedInline  on gm.code so we see what the graph looks like?

zou3519 reviewed (commented):
Deferring to @bdhirsh for the AOTAutograd-related pieces. The HOP-related pieces look reasonable. High-level comments:
- if we are modeling multiple effects (it seems like we are), we should have a mapping of effect to token instead of (op, torchbind object) -> token
- we should check the calling convention on the graph returned by AOTAutograd (e.g. it should have additional tokens) via self.assertExpectedInline

angelayi committed:
Update on "[hoo][wip] Add with_effects to handle side effectful ops"


Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Result with ScriptObject:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=2] = placeholder[target=arg1_1]
    %_tensor_constant0 : [num_users=1] = get_attr[target=_tensor_constant0]
    %with_effects : [num_users=2] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, _TorchScriptTesting.takes_foo.default, %_tensor_constant0, %arg1_1), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %getitem_1 : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 1), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %getitem_1), kwargs = {})
    return [getitem, add]
```

[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,198 @@
+
+
+def get_effect_key(op, args, kwargs) -> Optional[_EffectType]:
+    if op in SIDE_EFFECTFUL_OPS:
+        return _EffectType.ORDERED
```
SIDE_EFFECTFUL_OPS might be better as a mapping from op to _EffectType ? But this is fine for now since we only have one effect

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,198 @@
+) -> Any:
+    """
+    Args:
+        tracing: Whether or not we are currently tracing. If we are not tracing,
+        we will create tokens for every effect seen that does not already have a
+        token. If we ware tracing, then the `tokens` dict should already be
+        populated, and we will query tokens from there.
```
When are we not tracing?

angelayi replied:
I'm not sure whats a good name, but I want to distinguish between run_functionalized_fw_and_collect_metadata which just collects metadata (so i consider it "not tracing"), and then the make_fx dispatch tracing.

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -220,6 +220,21 @@ def __init__(self, pre_dispatch=False, export=False):
+        # tracing=False, we will count the number of tokens needed. In the
+        # second functionalization pass, when tracing=True, we will insert
+        # tokens into the graph.
+        self._tracing = _tracing
```
To clarify: we just need this flag to have a nice assertion, right?

zou3519 replied:
If we keep this, could we make the name a bit more descriptive? If someone just wants to use FunctionalTensorMode outside of AOTAutograd, it's not clear if they should set _tracing to be True or False. What the flag is really saying is something like "hey, you've got tokens, use these tokens"

bdhirsh replied:
Hmm I'm pretty neutral on the new argument. Alternatively, you could avoid adding the new state by querying whether there is currently a proxy mode active (using `torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY)`). This avoids the state, although it is definitely more less explicit (we're implicitly relying on the fact that aot autograd runs functionalization twice, once without proxy mode and once with proxy mode. Although this is pretty unlikely to change).

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,191 @@
+            gm.code
+        )
+
+        # With functionalization, it should appear wrapped with tokenize()
```
tokenize -> with_effects?

angelayi replied:
oops... previously i had named the op "tokenize" 😅

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,191 @@
+        inputs = (torch.randn(3),)
+
+        # With functionalization, it should appear wrapped with tokenize()
+        gm, gs = aot_export_module(M(), inputs, trace_joint=False)
```
I'm not familiar with this API -- what is gs?

angelayi replied:
graph signature -- it's this type: https://fburl.com/code/32p134uh

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,191 @@
+
+@unittest.skipIf(not torch._dynamo.is_dynamo_supported(), "dynamo isn't support")
+class TestWithEffects(TestCase):
+    def setUp(self):
+        if IS_MACOS:
+            raise unittest.SkipTest("non-portable load_library call used in test")
+        elif IS_SANDCASTLE or IS_FBCODE:
+            torch.ops.load_library(
+                "//caffe2/test/cpp/jit:test_custom_class_registrations"
+            )
+        elif IS_WINDOWS:
+            lib_file_path = find_library_location("torchbind_test.dll")
+            torch.ops.load_library(str(lib_file_path))
+        else:
+            lib_file_path = find_library_location("libtorchbind_test.so")
+            torch.ops.load_library(str(lib_file_path))
```
nit: Could we do something like mark the below tests with skipIf decorators if they use torchbind and torchbind wasn't found?

I'm not sure why, but I'm unable to actually find libtorchbind_test.so

angelayi replied:
I ended up skipping the test the entire test directly because I think I need to enable tokens on export before allowing torchbind ops 😅

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,191 @@
+            str(gm.code).strip(),
+            """\
+def forward(self, arg0_1, arg1_1):
+    with_effects = torch._higher_order_ops.effects.with_effects(arg0_1, torch.ops.aten._print.default, 'moo');  arg0_1 = None
```
(no action needed) I'm kinda curious what happens if someone puts a _print inside one of the cond branches. We would need to update cond to consume the token and output a token. cc @ydwu4

zou3519 commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -377,9 +377,10 @@ def create_graph_signature(
     graph_output_names = _graph_output_names(fx_g)
 
     num_params_buffers = len(param_names) + len(buffer_names)
+    num_tokens = len(fw_metadata.tokens)
```
Did you want to test multiple types of tokens or should we punt that? I'm fine with just punting it (to test multiple types you'd need to define another op that has a different side effect :P. Maybe "_print(str, ordered=True)" ?)

zou3519 replied:
It's good to test it now so that we don't specialize on num_tokens=1 in the followup work (export+inductor)

angelayi replied:
Does that mean I should create an "OrderedEffect" and "UnorderedEffect" where if print(...ordered=True) then it's an OrderedEffect otherwise it's UnorderedEffect?

zou3519 replied:
Er, maybe punt this until we have a use case. Any sort of "unordered effect" needs to be implemented slightly differently, that is, each call to print(ordered=False) needs to get its own token.

bdhirsh replied:
if we want to not bother with tests for now / punt, maybe add a comment / assert somewhere for now that `num_tokens == 1`?

zou3519 commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -350,12 +349,28 @@ def _functionalized_f_helper(*args):
         )
-        with disable_above, FunctionalTensorMode(aot_config.pre_dispatch):
+
+        with disable_above:
```
what happened here? it looks like we used to create a new FunctionalTensor mode but we no longer do?

angelayi replied:
Yeah, it seems to not be necessary as this call gets wrapped with [_create_graph](https://github.com/pytorch/pytorch/blob/3eefe962974be9cd70db7717894c010207f9d88b/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py#L34) which already has a FunctionalMode. 

Although I am breaking some tests.. not sure if this is related

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,191 @@
+        ):
+            make_fx(f)(torch.tensor([]), torch.tensor(4))
+
+    def test_compile_aot_eager(self):
+        def f(x):
+            torch.ops.aten._print("moo")
+            res = x + x
+            torch.ops.aten._print("moo")
+            return res
+
+        inputs = (torch.randn(2, 3),)
+
+        res = torch.compile(f, backend="aot_eager")(*inputs)
+        self.assertTrue(torch.allclose(res, f(*inputs)))
+
```
can you test a case where inputs has requires_grad=True?

zou3519 commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -470,6 +485,10 @@ def joint_helper(primals, tangents):
+    # Additionally pass in tokens as inputs
+    additional_token_inputs = [torch.tensor([])] * len(meta.tokens)
+    args = [*additional_token_inputs, *args]
+
```
What happens in the requires_grad=True case? We trace out a joint fwd_backward graph that has tokens, and then partition it? If so, can we add an assertion that all the tokens show up in the partitioned fwd graph?

bdhirsh replied:
related question: what happens if someone wants to use an op that requires an effect token in the backward (e.g. in a custom autograd.Function)?

I think we should either:
(1) explicitly support it in this PR (doable but kind of a pain, since the backward doesn't have a runtime wrapper today, and we also don't do any "analysis" on the backward to see if there are any ops that needs this handling)
(2) figure out how to explicitly error and wait for someone to run into it

angelayi replied:
When requires_Grad=True, right now I just mark the tokens as "forward outputs" that need to be partitioned in the forward graph and not the backwards graph. There are a couple of assertions checking that the number of input/output nodes of the graphs are consistent with the ViewAndMutationMetadata, so I think those would be the "explicit errors" one might see if there ends up being a token in the backwards graph?

zou3519 replied:
seems reasonable

zou3519 reviewed (commented):
This looks reasonable to me. I'll defer to @bdhirsh for review on the AOTAutograd pieces. Added some suggestions for testing and some questions. Also, the tests look like they're failing.

bdhirsh commented on torch/_functorch/_aot_autograd/schemas.py:
```diff
@@ -267,6 +267,11 @@ class ViewAndMutationMeta:
+    # Map of effect type (ex. _EffectType.ORDERED) to token.  If there are
+    # side-effectful operators, FunctionalTensorMode will populate this
+    # dictionary telling us how many tokens we will need during tracing.
+    tokens: Dict[Any, torch.Tensor] = field(default_factory=dict)
```
are the values here guaranteed to be FakeTensors? (and if so maybe we should make the type annotation `[Any, FakeTensor]`, to make it clear we aren't secretly holding on to tensor memory)

angelayi replied:
Hmm, I guess in the context of aot_autograd they're always FakeTensors?

angelayi replied:
Oh but sometimes theyre FunctionalTensors?

bdhirsh replied:
hmmm yeah fair, i guess we can't easily make the annotation more specific

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -69,10 +69,15 @@ def create_runtime_wrapper(
 
     def runtime_wrapper(*args):
+        # Pass in effect tokens
+        args = (*[torch.tensor([])] * num_tokens, *args)
```
It would be nice to have a `Note [Side-Effectful Tokens in AOTAutograd]` comment somewhere, where you can:

(1) describe a bit more what aot autograd does here (add extra "effect token" tensor args as inputs/outputs in the graph

(2) Link the comment in the various sections of AOTAutograd that require special handling, so someone can go back and easily find all the spots (the runtime section here, and in schemas.py / collect_metadata_analysis.py)

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -350,12 +349,28 @@ def _functionalized_f_helper(*args):
             # Wrap inputs into functional wrappers
             f_args = pytree.tree_map(to_fun, args)
+
+            # Populate the current FunctionalTensorMode with the tokens per
```
Maybe I am too big a fan of `Note [...]` comments...

Do you mind adding a comment like:
```
Note [FunctionalTensorMode is Stateful]
```

At the top of FunctionalTensorMode that briefly explains why it's stateful, and then link the comment here? (aka the one spot in the codebase today that needs to handle propagating the state)

bdhirsh commented on torch/_functorch/_aot_autograd/traced_function_transforms.py:
```diff
@@ -350,12 +349,28 @@ def _functionalized_f_helper(*args):
+
+            # Populate the current FunctionalTensorMode with the tokens per
+            # operator
+            tokens = f_args[: len(meta.tokens)]
```
One sanity assert you could do here is to assert that `[x.dim == 0 for x in tokens]` (a common "failure mode" in aot autograd historically is that we mess up our indexing somewhere and plumb the wrong tensors around)

bdhirsh commented:
Mostly left nits about comments, but AOTAutograd stuff mostly looks good to me! (minus we should maybe figure out if/how we want to error if there are side effect ops in the backward. I could maybe buy the argument that this is broken today anyway, so "fixing" effect ops in the forward but doing nothing in the backward is strictly better than the current state)

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,198 @@
+)
+
+
+SIDE_EFFECTFUL_OPS = {torch.ops.aten._print.default}
```
I'm not sure how to distinguish the two. This list seems to be for ops such that if you reorder them, some correctness issue will occur, but the FX list seems to be to prevent certain ops with users=0 from being DCE-d. I changed this list to a map instead though, hopefully that could help?

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -0,0 +1,208 @@
+)
+
+
+@unittest.skipIf(not torch._dynamo.is_dynamo_supported(), "dynamo isn't support")
```
does skipIfTorchDynamo work here?

zou3519 commented on torch/_higher_order_ops/effects.py:
```diff
@@ -0,0 +1,203 @@
+with_effects = WithEffects()
+
+
+def has_aliasing(op: torch._ops.OpOverload):
+    for arg in op._schema.arguments:
+        if arg.alias_info is not None:
+            return True
+    return False
```
We should also check op._schema.returns for alias_info

zou3519 commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -105,7 +105,8 @@ def inner(*flat_args):
         # It doesn't matter if we run this under predispatch or not because it is
         # only for figuring out metadata
-        with disable_above, FunctionalTensorMode():
+        mode = FunctionalTensorMode(_allow_token_discovery=False)
```
I'm confused: this is the place we're collecting metadata. Shouldn't we be allowing discovery of new tokens?

zou3519 commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -105,7 +105,8 @@ def inner(*flat_args):
         # It doesn't matter if we run this under predispatch or not because it is
         # only for figuring out metadata
-        with disable_above, FunctionalTensorMode():
+        mode = FunctionalTensorMode(_allow_token_discovery=False)
```
Nvm, my question still stands: I'm confused: this looks like the place we're collecting metadata. Shouldn't we be allowing discovery of new tokens?

angelayi replied:
Oh sorry i think i flopped the name/use of the token. 😂

zou3519 reviewed (approved):
LGTM. I was confused about one of the callsites, please read

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi committed:
Update on "[hoo] Add with_effects to handle side effectful ops"


Proposal: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bnm38nu3yfno
Implementation discussion: https://docs.google.com/document/d/179QyhicGzTXJ5jvTAoAosP_Nzgf3PpgZwU_E3VV9PlM/edit#heading=h.bj61609o1buq
    
Result with print:
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %with_effects : [num_users=1] = call_function[target=torch._higher_order_ops.effects.with_effects](args = (%arg0_1, aten.print.default, moo), kwargs = {})
    %getitem : [num_users=1] = call_function[target=operator.getitem](args = (%with_effects, 0), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg1_1, %arg1_1), kwargs = {})
    return [getitem, add]
```

Follow ups:
* Add handling to auto_functionalize 
* Add support for tokens on the export side
* Add support for tokens on the inductor side


[ghstack-poisoned]

angelayi commented:
@pytorchbot merge -i

angelayi commented:
@pytorchbot merge
